### PR TITLE
Support string-encoded date arguments in `add|subtract*()`

### PR DESCRIPTION
**Pull request by rschu1ze**
Needed for #54795

Reviewer note: only https://github.com/ClickHouse/ClickHouse/pull/55869/commits/c2215cb5f0647947395fe8363954f02205ea34eb is relevant, the other commits are refactorings

Support for date-encoded string arguments in the plus and minus operators is temporarily reverted, e.g. `SELECT '2023-10-22' + INTERVAL 1 DAY` (80a048d5f89d7b7a49463ceea41d05d745f33c89). Will push these as a separate PR as it triggers some test failures.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `(add|subtract)(Year|Quarter|Month|Week|Day|Hour|Minute|Second|Millisecond|Microsecond|Nanosecond)` now support string-encoded date arguments, e.g. `SELECT addDays('2023-10-22', 1)`. This increases compatibility with MySQL and is needed by Tableau Online.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 7ef161eacd699511bd08a908fc2c9005455cd1d0 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55869/7ef161eacd699511bd08a908fc2c9005455cd1d0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6587333549">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55869/7ef161eacd699511bd08a908fc2c9005455cd1d0/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Functions/FunctionDateOrDateTimeAddInterval.h [hunk 1]
```diff
@@ -437,55 +489,100 @@ struct Adder
...
+        else
+        {
+            auto & vec_to = col_to.getData();
+
+            size_t size = delta.size();
+            vec_to.resize(size);
+
+            for (size_t i = 0; i < size; ++i)
+                vec_to[i] = transform.execute(from, checkOverflow(delta.getData()[i]), time_zone, utc_time_zone, scale);
+        }
```
Comment (antonio2368) [inline src/Functions/FunctionDateOrDateTimeAddInterval.h:550]: Same chunk twice?

File: src/Functions/SubtractSubSeconds.cpp [hunk 2]
```diff
@@ -1,28 +0,0 @@
-#include <Functions/FunctionFactory.h>
-#include <Functions/FunctionDateOrDateTimeAddInterval.h>
-
-
-namespace DB
-{
-
-using FunctionSubtractNanoseconds = FunctionDateOrDateTimeAddInterval<SubtractNanosecondsImpl>;
-REGISTER_FUNCTION(SubtractNanoseconds)
```
Comment (antonio2368) [inline src/Functions/SubtractSubSeconds.cpp:9]: why was it split?

Comment (rschu1ze) [reply src/Functions/SubtractSubSeconds.cpp:9]: Just for consistency with existing files subtractDays.cpp, subtractHours.cpp, subtractMinutes.cpp, subtractSeconds.cpp etc. The split is part of a small refactoring unrelated to the actual changes (see the PR message).

File: docs/en/sql-reference/functions/date-time-functions.md [hunk 3]
```diff
@@ -2388,42 +2391,50 @@ Like function `YYYYMMDDhhmmssToDate()` but produces a [DateTime64](../../sql-ref
...
-└─────────────────────┴──────────────────────────┘
+┌─add_years_with_date─┬─add_years_with_date_time─┬─add_years_with_date_time_string─┐
+│          2025-01-01 │      2025-01-01 00:00:00 │         2025-01-01 00:00:00.000 │
+└─────────────────────┴──────────────────────────┴─────────────────────────────────┘
 ```
 
-## subtractYears, subtractMonths, subtractWeeks, subtractDays, subtractHours, subtractMinutes, subtractSeconds, subtractQuarters
+## subtractYears, subtractQuarters, subtractMonths, subtractWeeks, subtractDays, subtractHours, subtractMinutes, subtractSeconds, subtractMilliseconds, subtractMicroseconds, subtractNanoseconds
+
+These functions subtract units of the interval specified by the function name from a date, a date with time or a date or date with time encoded as string. A date or date with time is returned.
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply docs/en/sql-reference/functions/date-time-functions.md:2419]: Thanks, rephrased it a bit.

File: src/Functions/FunctionDateOrDateTimeAddInterval.h (hunk 1 above)
Comment (rschu1ze) [reply src/Functions/FunctionDateOrDateTimeAddInterval.h:550]: Removed the duplication, thanks.

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**